### DISCUSSION TITLE
Scale a deployment to 0 before sync

### DISCUSSION THREAD

**Discussion by JanCizmar** [argoproj/argo-cd#12551] [Q&A]
Hi there,

currently, I have this workflow for deploying my testing instance without ArgoCD (I am using bash script for that):

1. Scale the app deployment to 0
2. "copy" the production database to the testing one with pg_dump & pg_restore
3. update the image in used in the app deployment to new version
4. Scale the app deployment back to the desired replica count

If I don't scale the deployment to 0, it breaks Since the Apps database is truncated. The pods are restarted and run the database migrations automatically from scratch, which breaks the database copy operation. So I really need to stop the deployment.

Now I am trying to migrate this process to ArgoCD. So I am committing new image version to the my Argo's project repo. I would like to create a pre-sync hook which would do the database copy operation. But I don't know how to do the scaling. One option is to scale the deployment using kubectl manually in the job, but maybe there is another better way how to do that, which I am missing.

Maybe it's possible with [Sync Phases and Waves](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/#sync-phases-and-waves)?

Thanks!

**Comment by kostis-codefresh** [comment]
One of the main selling points of Kubernetes is zero downtime deployments.

Scaling to zero during a deployment goes against this advantage.

In general, pods should not migrate db during startup. You should decouple database migration from application deployment. Use a Kubernetes job or an external CI system. Database upgrades should have their own lifecycle.

**Reply by JanCizmar** [reply]
> Database upgrades should have their own lifecycle.

I am using Spring Boot + Liquibase, where migrating on startup is pretty common scenario. I don't see any reason apart from my specific use case, why it should be decoupled.

Anyway, the DB migration is not only reason why It's good idea to stop the deployment. It's a testing environment, so there is no need for 0 downtime. But it's not good idea to have the deployment running while the database is being deleted by other task, the deployed app shouldn't try to modify the objects in the DB.

**Reply by kostis-codefresh** [reply]
This is probably unrelated to ArgoCD but there is a lot of material already on why applications should not do migrations themselves. It makes rollbacks super difficult and it opens several issues with concurrency. Example https://pythonspeed.com/articles/schema-migrations-server-startup/

>migrating on startup is pretty common scenario

That was a pretty common scenario with VMs/baremetal/monoliths. Doesn't mean it is a good idea with Kubernetes/ArgoCD



Why do you need to delete a db everytime a deployment happens? Is this a common scenario or just an edge case?

**Reply by JanCizmar** [reply]
> Example https://pythonspeed.com/articles/schema-migrations-server-startup/

Good article thanks. 

> Why do you need to delete a db everytime a deployment happens? 

I need to copy the production database there and test whether the app works after the migration (for example).

> Is this a common scenario or just an edge case?

Not sure.

**Reply by kostis-codefresh** [reply]
I am not a posgresql expert, but what is wrong with the following scenario?

1. Prod app is running with prod db, staging is running with staging db. 
2. Use pg_dump & pg_restore in a pre-sync hook (but only for data) to copy prod db to staging db. Staging app should still run at this point
3. Run db migrations in another pre-sync hook. If there are errors the whole process stops and you should fix the db migration scripts
4. Main sync phase runs and staging goes to next version

Deployment is finished. New version is in staging with new schema/migrations

Zero downtime, and no need to scale down anything. 

This wouldn't work in your case? Am I missing something?

**Reply by JanCizmar** [reply]
> Use pg_dump & pg_restore in a pre-sync hook (but only for data) to copy prod db to staging db. Staging app should still run at this point

I have to pg_dump and pg_restore also the schema, since the prod can be behind the staging in migrations. In that case the it may fail.

But it looks like I am going to solve it by revoking the apps permission to connect, doing the dump and restore and then enabling it to connect again.

I don't see much benefit from decoupling the migrations from the app, since liquibase solves the issues with possible paralel runs (by locking), so theres not much benefit in that.

**Reply by JanCizmar** [reply]
The issue is that the app pod is restarted since it's unhealthy and it runs the migrations, which breaks the pg_restore. I have 2 options how to solve it.
1. Decouple the migrations, so the it won't run the migrations
2. Disable the app from running the migrations by scaling it to 0 
3.  Disable the app from running the migrations by revoking it's connect permissions

I preferred the 2nd option, so the app won't throw any errors. I don't feel like destroying database of live app is correct solution, which happens in both 1 and 3, but since it's a staging environment, its OK.

**Reply by kostis-codefresh** [reply]
Decoupling migrations is my recommendation.
